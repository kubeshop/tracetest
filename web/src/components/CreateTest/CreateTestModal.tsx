import {useRef} from 'react';
import {Modal, Form, Input, Button, Select, Checkbox} from 'antd';
import {DeleteOutlined} from '@ant-design/icons';
import {useCreateTestMutation} from 'services/TestService';
import {HTTP_METHOD} from 'types';
import './CreateTest.css';

interface IProps {
  visible: boolean;
  onClose: () => void;
}

const CreateTestModal = ({visible, onClose}: IProps): JSX.Element => {
  const [createTest, result] = useCreateTestMutation();
  const [form] = Form.useForm();
  const touchedHttpHeadersRef = useRef<{[key: string]: Boolean}>({});
  const onFinish = (values: any) => {
    const headers = values.headersList
      .filter((i: {checked: boolean}) => i.checked)
      .map(({key, value}: {key: string; value: string}) => ({key, value}));
    createTest({
      name: values.name,
      serviceUnderTest: {
        request: {url: values.url, method: values.method, body: values.body, headers},
      },
    });
    onClose();
  };

  const onFinishFailed = () => {};

  const renderActionButtons = () => {
    return (
      <>
        <Button type="ghost" htmlType="button" onClick={onClose}>
          Cancel
        </Button>

        <Button type="primary" form="newTest" htmlType="submit">
          Create Test
        </Button>
      </>
    );
  };

  return (
    <Modal
      closable
      title="Create New Test"
      visible={visible}
      footer={renderActionButtons()}
      onCancel={onClose}
      wrapClassName="test-modal"
    >
      <div style={{display: 'flex'}}>
        <Form
          name="newTest"
          form={form}
          initialValues={{remember: true}}
          onFinish={onFinish}
          onFinishFailed={onFinishFailed}
          autoComplete="off"
          layout="vertical"
        >
          <div style={{display: 'flex', marginBottom: 24}}>
            <Form.Item name="method" initialValue="GET" valuePropName="value" noStyle>
              <Select style={{minWidth: 120}} className="method-select" dropdownClassName="method-select-item">
                {Object.keys(HTTP_METHOD).map(el => {
                  return (
                    <Select.Option key={el} value={el}>
                      {el}
                    </Select.Option>
                  );
                })}
              </Select>
            </Form.Item>

            <Form.Item name="url" rules={[{required: true, message: 'Please input Endpoint!'}]} noStyle>
              <Input placeholder="Enter request url" />
            </Form.Item>
          </div>

          <Form.Item
            name="name"
            label="Name"
            colon={false}
            wrapperCol={{span: 24, offset: 0}}
            rules={[{required: true, message: 'Please input test name!'}]}
          >
            <Input />
          </Form.Item>

          <Form.Item label="Headers List" wrapperCol={{span: 24, offset: 0}}>
            <div style={{minHeight: 80}}>
              <Form.List name="headersList" initialValue={[{}, {}, {}]}>
                {(fields, {add, remove}) => (
                  <>
                    {fields.map((field, index) => (
                      <div key={field.name} style={{display: 'flex', alignItems: 'center'}}>
                        <Form.Item name={[field.name, 'checked']} valuePropName="checked" noStyle>
                          <Checkbox style={{marginRight: 8}} />
                        </Form.Item>

                        <Form.Item name={[field.name, 'key']} noStyle>
                          <Input
                            placeholder={`Header${index + 1}`}
                            onChangeCapture={() => {
                              if (!touchedHttpHeadersRef.current[field.name]) {
                                touchedHttpHeadersRef.current[field.name] = true;
                                const headers = form.getFieldsValue().headersList;
                                headers[field.name].checked = true;
                                form.setFieldsValue({headersList: headers});
                              }

                              if (fields.length - 1 === index) {
                                add({checked: false});
                              }
                            }}
                          />
                        </Form.Item>
                        <Form.Item noStyle name={[field.name, 'value']}>
                          <Input
                            placeholder={`Value${index + 1}`}
                            onChangeCapture={() => {
                              if (!touchedHttpHeadersRef.current[field.name]) {
                                touchedHttpHeadersRef.current[field.name] = true;
                                const headers = form.getFieldsValue().headersList;
                                headers[field.name].checked = true;
                                form.setFieldsValue({headersList: headers});
                              }

                              if (fields.length - 1 === index) {
                                add({checked: false});
                              }
                            }}
                          />
                        </Form.Item>

                        <Form.Item noStyle>
                          <Button
                            style={{marginLeft: 8}}
                            type="text"
                            icon={<DeleteOutlined style={{fontSize: 24, color: '#D9D9D9'}} />}
                            onClick={() => {
                              touchedHttpHeadersRef.current[field.name] = false;
                              remove(index);
                              if (fields.length === 1 || fields.length - 1 === index) {
                                add();
                              }
                            }}
                          />
                        </Form.Item>
                      </div>
                    ))}
                  </>
                )}
              </Form.List>
            </div>
          </Form.Item>
          <Form.Item label="Request body" name="body" colon={false} wrapperCol={{span: 24, offset: 0}}>
            <Input.TextArea style={{maxHeight: 150, height: 120}} />
          </Form.Item>
        </Form>
      </div>
    </Modal>
  );
};

export default CreateTestModal;
