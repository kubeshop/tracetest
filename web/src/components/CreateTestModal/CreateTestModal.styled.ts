import {Typography} from 'antd';
import styled from 'styled-components';

export const Wrapper = styled.div`
  .test-modal {
    overflow: hidden !important;
    display: flex;
  }

  .test-modal .ant-modal {
    top: 50px;
  }

  .test-modal .ant-modal-content {
    height: min(calc(100vh - 16%), 620px);
  }

  .test-modal .ant-modal-body {
    position: absolute;
    min-height: fit-content;
    height: calc(100% - 110px);
    width: 100%;
    overflow-y: auto;
  }

  .test-modal .ant-modal-footer {
    position: absolute;
    left: 0;
    right: 0;
    bottom: 0;
  }

  .method-select .ant-select-selector {
    background-color: #fafafa !important;
  }

  .method-select-item .ant-select-item-option-selected {
    background-color: #fafafa !important;
  }
`;

export const DropdownText = styled(Typography.Text).attrs({
  as: 'a',
})``;

export const DemoTextContainer = styled.div`
  margin-bottom: 24px;
`;
