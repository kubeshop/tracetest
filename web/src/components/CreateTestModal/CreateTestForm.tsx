import {DeleteOutlined, DownOutlined, PlusOutlined} from '@ant-design/icons';
import {Button, Dropdown, Form, FormInstance, Input, Menu, Select, Typography} from 'antd';

import {Steps} from 'components/GuidedTour/homeStepList';
import {HTTP_METHOD} from 'constants/Common.constants';
import {DEFAULT_HEADERS, DemoTestExampleList} from 'constants/Test.constants';
import {camelCase} from 'lodash';
import React from 'react';
import GuidedTourService, {GuidedTours} from 'services/GuidedTour.service';
import Validator from 'utils/Validator';
import {TooltipQuestion} from '../TooltipQuestion/TooltipQuestion';
import * as S from './CreateTestModal.styled';
import CreateTestAnalyticsService from '../../services/Analytics/CreateTestAnalytics.service';

export const FORM_ID = 'create-test';

export interface ICreateTestValues {
  body: string;
  headers: {
    key: string;
    value: string;
  }[];
  method: HTTP_METHOD;
  name: string;
  url: string;
}

interface IProps {
  form: FormInstance<ICreateTestValues>;
  onSelectDemo(value: string): void;
  onSubmit(values: ICreateTestValues): Promise<void>;
  onValidation(isValid: boolean): void;
  selectedDemo: string;
}

const CreateTestForm = ({form, onSelectDemo, onSubmit, onValidation, selectedDemo}: IProps) => {
  const handleOnDemoClick = ({key}: {key: string}) => {
    CreateTestAnalyticsService.onDemoTestClick();
    onSelectDemo(key);
    const {body, description, method, name, url} = DemoTestExampleList.find(demo => demo.name === key) || {};

    form.setFieldsValue({
      body,
      method,
      name: `${name} - ${description}`,
      url,
    });

    onValidation(true);
  };

  const handleOnValuesChange = (changedValues: any, allValues: ICreateTestValues) => {
    const isValid =
      Validator.required(allValues.name) && Validator.required(allValues.url) && Validator.url(allValues.url);
    onValidation(isValid);
  };

  const menuLayout = (
    <Menu
      items={DemoTestExampleList.map(({name}) => ({
        key: name,
        label: <span data-cy={`demo-example-${camelCase(name)}`}>{name}</span>,
      }))}
      onClick={handleOnDemoClick}
      style={{width: '190px'}}
    />
  );

  return (
    <Form
      autoComplete="off"
      data-cy="create-test-modal"
      form={form}
      layout="vertical"
      name={FORM_ID}
      onFinish={onSubmit}
      onValuesChange={handleOnValuesChange}
    >
      <S.GlobalStyle />

      <S.DemoContainer>
        <Typography.Text>Try these examples in our demo env: </Typography.Text>
        <Dropdown overlay={menuLayout} placement="bottomLeft" trigger={['click']}>
          <Typography.Link data-cy="example-button" onClick={e => e.preventDefault()} strong>
            {selectedDemo || 'Choose Example'} <DownOutlined />
          </Typography.Link>
        </Dropdown>
        <TooltipQuestion
          margin={8}
          title="We have a microservice based on the Pokemon API installed - pick an example to autofill this form"
        />
      </S.DemoContainer>

      <S.Row>
        <div>
          <Form.Item name="method" initialValue={HTTP_METHOD.GET} valuePropName="value" noStyle>
            <Select
              className="select-method"
              data-cy="method-select"
              data-tour={GuidedTourService.getStep(GuidedTours.Home, Steps.Method)}
              dropdownClassName="select-dropdown-method"
              style={{minWidth: 120}}
            >
              {Object.keys(HTTP_METHOD).map(method => {
                return (
                  <Select.Option data-cy={`method-select-option-${method}`} key={method} value={method}>
                    {method}
                  </Select.Option>
                );
              })}
            </Select>
          </Form.Item>
        </div>

        <Form.Item
          name="url"
          rules={[
            {required: true, message: 'Please enter a request url'},
            {type: 'url', message: 'Request url is not valid'},
          ]}
          style={{flex: 1}}
        >
          <Input
            data-cy="url"
            data-tour={GuidedTourService.getStep(GuidedTours.Home, Steps.Url)}
            placeholder="Enter request url"
          />
        </Form.Item>
      </S.Row>

      <Form.Item
        className="input-name"
        data-cy="name"
        data-tour={GuidedTourService.getStep(GuidedTours.Home, Steps.Name)}
        label="Name"
        name="name"
        rules={[{required: true, message: 'Please enter a test name'}]}
      >
        <Input placeholder="Enter test name" />
      </Form.Item>

      <Form.Item
        className="input-headers"
        data-tour={GuidedTourService.getStep(GuidedTours.Home, Steps.Headers)}
        label="Headers list"
      >
        <Form.List name="headers" initialValue={DEFAULT_HEADERS}>
          {(fields, {add, remove}) => (
            <>
              {fields.map((field, index) => (
                <S.HeaderContainer key={field.name}>
                  <Form.Item name={[field.name, 'key']} noStyle>
                    <Input placeholder={`Header ${index + 1}`} />
                  </Form.Item>

                  <Form.Item name={[field.name, 'value']} noStyle>
                    <Input placeholder={`Value ${index + 1}`} />
                  </Form.Item>

                  <Form.Item noStyle>
                    <Button
                      icon={<DeleteOutlined style={{fontSize: 12, color: 'rgba(3, 24, 73, 0.5)'}} />}
                      onClick={() => remove(field.name)}
                      style={{marginLeft: 12}}
                      type="link"
                    />
                  </Form.Item>
                </S.HeaderContainer>
              ))}

              <Button
                data-cy="add-header"
                icon={<PlusOutlined />}
                onClick={() => add()}
                style={{fontWeight: 600, height: 'auto', padding: 0}}
                type="link"
              >
                Add Header
              </Button>
            </>
          )}
        </Form.List>
      </Form.Item>

      <Form.Item
        className="input-body"
        data-cy="body"
        data-tour={GuidedTourService.getStep(GuidedTours.Home, Steps.Body)}
        label="Request body"
        name="body"
        style={{marginBottom: 0}}
      >
        <Input.TextArea placeholder="Enter request body text" />
      </Form.Item>
    </Form>
  );
};

export default CreateTestForm;
