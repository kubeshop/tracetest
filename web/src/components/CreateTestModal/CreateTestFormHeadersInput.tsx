import {DeleteOutlined, PlusOutlined} from '@ant-design/icons';
import {Button, Form, Input} from 'antd';
import React from 'react';
import {DEFAULT_HEADERS} from '../../constants/Test.constants';
import GuidedTourService, {GuidedTours} from '../../services/GuidedTour.service';
import {Steps} from '../GuidedTour/homeStepList';
import * as S from './CreateTestModal.styled';

export const CreateTestFormHeadersInput: React.FC = () => (
  <Form.Item
    className="input-headers"
    data-tour={GuidedTourService.getStep(GuidedTours.Home, Steps.Headers)}
    label="Headers list"
  >
    <Form.List name="headers" initialValue={DEFAULT_HEADERS}>
      {(fields, {add, remove}) => (
        <>
          {fields.map((field, index) => (
            <S.HeaderContainer key={field.name}>
              <Form.Item name={[field.name, 'key']} noStyle>
                <Input placeholder={`Header ${index + 1}`} />
              </Form.Item>

              <Form.Item name={[field.name, 'value']} noStyle>
                <Input placeholder={`Value ${index + 1}`} />
              </Form.Item>

              <Form.Item noStyle>
                <Button
                  icon={<DeleteOutlined style={{fontSize: 12, color: 'rgba(3, 24, 73, 0.5)'}} />}
                  onClick={() => remove(field.name)}
                  style={{marginLeft: 12}}
                  type="link"
                />
              </Form.Item>
            </S.HeaderContainer>
          ))}

          <Button
            data-cy="add-header"
            icon={<PlusOutlined />}
            onClick={() => add()}
            style={{fontWeight: 600, height: 'auto', padding: 0}}
            type="link"
          >
            Add Header
          </Button>
        </>
      )}
    </Form.List>
  </Form.Item>
);
