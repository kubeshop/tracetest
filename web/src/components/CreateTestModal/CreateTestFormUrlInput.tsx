import {Form, Input, Select} from 'antd';
import React from 'react';
import {HTTP_METHOD} from '../../constants/Common.constants';
import GuidedTourService, {GuidedTours} from '../../services/GuidedTour.service';
import {Steps} from '../GuidedTour/homeStepList';
import * as S from './CreateTestModal.styled';

export const CreateTestFormUrlInput: React.FC = () => {
  return (
    <S.Row>
      <div>
        <Form.Item name="method" initialValue={HTTP_METHOD.GET} valuePropName="value" noStyle>
          <Select
            className="select-method"
            data-cy="method-select"
            data-tour={GuidedTourService.getStep(GuidedTours.Home, Steps.Method)}
            dropdownClassName="select-dropdown-method"
            style={{minWidth: 120}}
          >
            {Object.keys(HTTP_METHOD).map(method => {
              return (
                <Select.Option data-cy={`method-select-option-${method}`} key={method} value={method}>
                  {method}
                </Select.Option>
              );
            })}
          </Select>
        </Form.Item>
      </div>

      <Form.Item
        name="url"
        rules={[
          {required: true, message: 'Please enter a request url'},
          {type: 'url', message: 'Request url is not valid'},
        ]}
        style={{flex: 1}}
      >
        <Input
          data-cy="url"
          data-tour={GuidedTourService.getStep(GuidedTours.Home, Steps.Url)}
          placeholder="Enter request url"
        />
      </Form.Item>
    </S.Row>
  );
};
