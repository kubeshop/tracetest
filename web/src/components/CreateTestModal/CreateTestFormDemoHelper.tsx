import {DownOutlined} from '@ant-design/icons';
import {Dropdown, FormInstance, Menu, Typography} from 'antd';
import {camelCase} from 'lodash';
import React from 'react';
import {DemoTestExampleList} from '../../constants/Test.constants';
import CreateTestAnalyticsService from '../../services/Analytics/CreateTestAnalytics.service';
import {TooltipQuestion} from '../TooltipQuestion/TooltipQuestion';
import {ICreateTestValues} from './CreateTestForm';
import * as S from './CreateTestModal.styled';

interface IProps {
  form: FormInstance<ICreateTestValues>;
  onSelectDemo(value: string): void;
  onValidation(isValid: boolean): void;
  selectedDemo: string;
}

export const CreateTestFormDemoHelper: React.FC<IProps> = ({selectedDemo, onSelectDemo, onValidation, form}) => {
  const handleOnDemoClick = ({key}: {key: string}) => {
    CreateTestAnalyticsService.onDemoTestClick();
    onSelectDemo(key);
    const {body, description, method, name, url} = DemoTestExampleList.find(demo => demo.name === key) || {};

    form.setFieldsValue({
      body,
      method,
      name: `${name} - ${description}`,
      url,
    });

    onValidation(true);
  };
  return (
    <S.DemoContainer>
      <Typography.Text>Try these examples in our demo env: </Typography.Text>
      <Dropdown
        overlay={() => (
          <Menu
            items={DemoTestExampleList.map(({name}) => ({
              key: name,
              label: <span data-cy={`demo-example-${camelCase(name)}`}>{name}</span>,
            }))}
            onClick={handleOnDemoClick}
            style={{width: '190px'}}
          />
        )}
        placement="bottomLeft"
        trigger={['click']}
      >
        <Typography.Link data-cy="example-button" onClick={e => e.preventDefault()} strong>
          {selectedDemo || 'Choose Example'} <DownOutlined />
        </Typography.Link>
      </Dropdown>
      <TooltipQuestion
        margin={8}
        title="We have a microservice based on the Pokemon API installed - pick an example to autofill this form"
      />
    </S.DemoContainer>
  );
};
