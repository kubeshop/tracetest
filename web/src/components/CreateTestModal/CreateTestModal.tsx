import {useCallback} from 'react';
import {useTour} from '@reactour/tour';
import {Modal, Form, Button} from 'antd';
import {useCreateTestMutation, useRunTestMutation} from 'redux/apis/Test.api';
import './CreateTestModal.styled.ts';
import {useNavigate} from 'react-router-dom';
import GuidedTourService, {GuidedTours} from '../../services/GuidedTour.service';
import {Steps} from '../GuidedTour/homeStepList';
import CreateTestAnalyticsService from '../../services/Analytics/CreateTestAnalytics.service';
import * as S from './CreateTestModal.styled';
import CreateTestForm, {ICreateTestValues} from './CreateTestForm';

interface IProps {
  visible: boolean;
  onClose: () => void;
}

const {onCreateTestFormSubmit} = CreateTestAnalyticsService;

const CreateTestModal = ({visible, onClose}: IProps): JSX.Element => {
  const navigate = useNavigate();
  const {setIsOpen} = useTour();
  const [createTest, {isLoading: isLoadingCreateTest}] = useCreateTestMutation();
  const [runTest, {isLoading: isLoadingRunTest}] = useRunTestMutation();

  const [form] = Form.useForm<ICreateTestValues>();

  const onCreate = useCallback(
    async (values: ICreateTestValues) => {
      const headers = values.headersList
        .filter((i: {checked: boolean}) => i.checked)
        .map(({key, value}: {key: string; value: string}) => ({key, value}));
      const newTest = await createTest({
        name: values.name,
        serviceUnderTest: {
          request: {url: values.url, method: values.method, body: values.body, headers},
        },
      }).unwrap();
      const newTestRunResult = await runTest(newTest.testId).unwrap();
      onClose();
      setIsOpen(false);
      navigate(`/test/${newTest.testId}?resultId=${newTestRunResult.resultId}`);
    },
    [createTest, navigate, onClose, runTest, setIsOpen]
  );

  const renderActionButtons = () => {
    return (
      <>
        <Button type="ghost" htmlType="button" onClick={onClose}>
          Cancel
        </Button>

        <Button
          type="primary"
          form="newTest"
          onClick={() => {
            onCreateTestFormSubmit();
            form.submit();
          }}
          loading={isLoadingCreateTest || isLoadingRunTest}
          data-tour={GuidedTourService.getStep(GuidedTours.Home, Steps.Run)}
          data-cy="create-test-submit"
        >
          Run Test
        </Button>
      </>
    );
  };

  return (
    <S.Wrapper>
      <Modal
        closable
        title="Create New Test"
        visible={visible}
        footer={renderActionButtons()}
        onCancel={onClose}
        wrapClassName="test-modal"
      >
        <div style={{display: 'flex'}} data-cy="create-test-modal">
          <CreateTestForm onSubmit={onCreate} form={form} />
        </div>
      </Modal>
    </S.Wrapper>
  );
};

export default CreateTestModal;
