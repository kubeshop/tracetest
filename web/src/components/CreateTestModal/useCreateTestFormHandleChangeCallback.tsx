import {useCallback} from 'react';
import Validator from '../../utils/Validator';
import {ICreateTestValues} from './CreateTestForm';

export const useCreateTestFormHandleChangeCallback = (
  onValidation: (isValid: boolean) => void
): ((changedValues: any, allValues: ICreateTestValues) => void) => {
  const authValidation = (testValues: ICreateTestValues): boolean => {
    switch (testValues.auth?.type) {
      case 'apiKey':
        return Validator.required(testValues.auth?.apiKey?.key) && Validator.required(testValues.auth?.apiKey?.value);
      case 'basic':
        return (
          Validator.required(testValues.auth?.basic?.username) && Validator.required(testValues.auth?.basic?.password)
        );
      case 'bearer':
        return Validator.required(testValues.auth?.bearer?.token);
      default:
        return true;
    }
  };
  return useCallback(
    (changedValues: any, allValues: ICreateTestValues) => {
      const isValid =
        Validator.required(allValues.name) &&
        Validator.required(allValues.url) &&
        Validator.url(allValues.url) &&
        authValidation(allValues);
      onValidation(isValid);
    },
    [onValidation]
  );
};
