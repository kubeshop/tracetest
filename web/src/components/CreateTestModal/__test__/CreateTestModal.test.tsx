import {render} from '@testing-library/react';
import {MemoryRouter} from 'react-router-dom';
import CreateTestModal from '../CreateTestModal';
import {ReduxWrapperProvider} from '../../../redux/ReduxWrapperProvider';

test('CreateTestModal', () => {
  const result = render(
    <MemoryRouter>
      <ReduxWrapperProvider>
        <CreateTestModal visible onClose={jest.fn()} />
      </ReduxWrapperProvider>
    </MemoryRouter>
  );
  expect(result.container).toMatchSnapshot();
});
