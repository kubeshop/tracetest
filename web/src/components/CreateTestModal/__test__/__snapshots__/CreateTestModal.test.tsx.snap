// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateTestModal 1`] = `
<div>
  <div
    class="sc-bdvvtL gmMIqV"
  />
</div>
`;
